In [1]:
from jupyter_dash import JupyterDash
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Output, Input
import os 
import pandas as pd
import numpy as np
from plotly import graph_objects as go

In [2]:
pwd

'c:\\Users\\HP\\source\\repos\\STARS_dashboard\\stars-dashboard'

In [3]:
os.chdir(r"C:\Users\HP\source\repos\STARS_dashboard")
os.getcwd()

'C:\\Users\\HP\\source\\repos\\STARS_dashboard'

In [2]:
# app instantiation
app = JupyterDash(__name__)

In [4]:
# app layout
app.layout = html.Div([
    dcc.Dropdown(
        options=[{"label" : color,
                 "value" : color }
                 for color in ["blue","green","yellow"]
    ]),
    html.Div()
])

In [ ]:
# run the app
if __name__ == "__main__":
    app.run_server(mode = "inline",port = "3333")

In [ ]:
#using callbacks

# app instantiation
app = JupyterDash(__name__)

# app layout
app.layout = html.Div([
    dcc.Dropdown(id= "color_dropdown",
                 options = [
                    {"label": color, 
                    "value": color}
                    for color in ["Blue","Black", "Green"]
                ]),
    html.Br(),
    html.Div(id="display_dropdown")
])

# callback functions
@app.callback(Output(component_id = "display_dropdown",
                     component_property= "children"),
              Input(component_id = "color_dropdown",
                    component_property = "value"))

def display_selected_color(color):
    if color is None:
        color = "Nothing"
    return "The color is " + color 

# run the app
if __name__ == "__main__":
    app.run_server(mode = "inline", port = 2)

In [ ]:
# app instantiation
app = JupyterDash(__name__)

# information on the activities
timeline = ["first year", "second year"]
short_description = {"first year":"The data collection activities spanned 10 districts while gathering data on 343 schools where 100 schools were in the\
                     control group while 243 schools were in the treatment group",
                     "second year":"489 schools were visited across 10 districts to conduct STARS data collection activities."}

# app layout
app.layout = html.Div([
    dcc.Dropdown(id = "id_dropdown",
                 options = [{"label" : time,
                            "value": time}
                            for time in timeline]),
    html.Br(),
    html.Div(id = "output_dropdown")
])

# callback functions
@app.callback(Output(component_id = "output_dropdown",
                     component_property= "children"),
              Input(component_id = "id_dropdown",
                    component_property = "value"))

def activity_info(timeline):
    if timeline is None:
        return "STARS project is being implemented by Innovation for Poverty Actions in partnership with Georgetown University, MINEDUC, REB, and NESA"
    return [html.H3(timeline), f'For the  {timeline}, {short_description[timeline]}']

# run the app
if __name__ == "__main__":
    app.run_server(mode = "inline", port = 5)

## Working with Plotly's Figure Objects

### Understanding the Figure Object

In [ ]:
go.Figure()

In [ ]:
fig = go.Figure()
fig.add_scatter(x=[1,2,3],y=[4,2,3])
fig.show()

In [ ]:
fig.add_scatter(x = [3,5,2,4], y = [5,2,6,0])

#### Layout Attribute

In [ ]:
fig.layout.title = "Example 0"
fig.layout.xaxis.title = "x axis"
fig.layout.yaxis.title = "y axis"
fig.show()

In [ ]:
# exploring figure objects
fig.show("json")

In [ ]:
fig.show()

In [ ]:
fig.show(config = {"displaylogo" : False,
                   "modeBarButtonsToAdd" : ["drawrect",
                                            "drawcircle",
                                            "eraseshape"]})

#### Converting figures

In [9]:
# To Html
fig.write_html("html_plot.html",
               config = {"toImageButtonOptions" : {"format" : "svg"}})

In [ ]:
# to images
fig.write_image("apps/images/example_image.svg",
                height = 800, width = 900)

# plotting using real data

In [4]:
# import datasets to use

# we start by using data entry data from year 1

data_entry_team_1 = pd.read_stata(r"data\ipa-data\learning_assessment\data_entry\year1\clean\first_entry\Final Assessment First Entry Clean.dta")
data_entry_team_2 = pd.read_stata(r"data\ipa-data\learning_assessment\data_entry\year1\clean\second_entry\Final Assessment Second Entry Clean.dta")
data_entry_reconciliation = pd.read_stata(r"data\ipa-data\learning_assessment\data_entry\year1\clean\reconciliation\reconciled_clean.dta")

In [5]:
data_entry_team_1.head(5)

deviceid devicephonenum  \
0  E0D3C54C-EFFE-4185-8F00-80D1CE7B79E6                  
1                      e9129293624d72e5                  
2                      d73df25fad2cc33b                  
3                      d6ee4c5c7759f9a6                  
4                      87c4ab681ae1d513                  

                          username  \
0                          collect   
1                          collect   
2                          collect   
3                          collect   
4  collect (not yet authenticated)   

                                         device_info duration caseid  \
0  Apple|iPhone|15.7.5|SurveyCTO Collect 2.80 (27...      180          
1  samsung|SM-T225|13|SurveyCTO Collect 2.80.2 (2...      217          
2  samsung|SM-T225|13|SurveyCTO Collect 2.80.2 (2...      575          
3  samsung|SM-T290|11|SurveyCTO Collect 2.80 (9b6...      401          
4  samsung|SM-T505N|10|SurveyCTO Collect 2.80.2 (...      429          

                    spv                  fo district    sector  ...  \
0       Shema Christian  Ndagijimana Samuel   Nyanza    Busoro  ...   
1  NAHIMANA Reverien      NDABANANIYE David  Musanze    Busogo  ...   
2       Shema Christian  Habonimana Gabriel  Gakenke  Kivuruga  ...   
3    NTAREMBA George       IRASUBIZA Zawadi  Musanze    Gacaca  ...   
4  NAHIMANA Reverien        KWIHANGANA Marc   Nyanza    Busoro  ...   

  kin1marks_p6_kin1p6_kin_q061p6_k kin1marks_p6_kin1p6_kin_q101p6_k v453  \
0                              NaN                              NaN  NaN   
1                              NaN                              NaN  NaN   
2                              NaN                              NaN  NaN   
3                              NaN                              NaN  NaN   
4                              NaN                              NaN  NaN   

  kin1marks_p5_kin1p5_kin_q141p5_k      submissiondate           starttime  \
0                              NaN 2023-08-07 17:02:03 2023-08-07 15:07:01   
1                              NaN 2023-07-28 16:33:29 2023-07-28 15:28:16   
2                              NaN 2023-07-24 15:34:11 2023-07-24 09:05:49   
3                              NaN 2023-07-31 11:31:03 2023-07-31 10:53:15   
4                              NaN 2023-08-07 09:32:26 2023-08-07 09:18:01   

              endtime    subdate    enddate v454  
0 2023-08-07 15:10:01 2023-08-07 2023-08-07  NaN  
1 2023-07-28 15:31:52 2023-07-28 2023-07-28  NaN  
2 2023-07-24 09:15:25 2023-07-24 2023-07-24  NaN  
3 2023-07-31 10:59:56 2023-07-31 2023-07-31  NaN  
4 2023-08-07 09:25:10 2023-08-07 2023-08-07  NaN  

[5 rows x 456 columns]

In [6]:
data_entry_team_1.dropna(axis=1,how="all").describe(include = [np.number]).T

count          mean           std           min  \
p1_leg_cwpm_sub1   1667.0  1.094301e+01  9.928665e+00  0.000000e+00   
p2_leg_cwpm_sub1   1675.0  2.356716e+01  1.297464e+01  0.000000e+00   
p3_leg_cwpm_sub1   1651.0  3.586978e+01  1.485797e+01  0.000000e+00   
formdef_version   56906.0  2.402470e+09  2.122712e+07  2.307212e+09   

                           25%           50%           75%           max  
p1_leg_cwpm_sub1  0.000000e+00  1.000000e+01  2.000000e+01  2.700000e+01  
p2_leg_cwpm_sub1  1.500000e+01  2.500000e+01  3.500000e+01  4.100000e+01  
p3_leg_cwpm_sub1  2.700000e+01  3.700000e+01  4.700000e+01  5.900000e+01  
formdef_version   2.407201e+09  2.407201e+09  2.407201e+09  2.407201e+09

In [7]:
data_entry_team_1.dropna(axis=1,how="all").describe(include="all",datetime_is_numeric=True).T

count unique  \
deviceid        56906    169   
devicephonenum  56906      5   
username        56906      5   
device_info     56906     57   
duration        56906   1456   
...               ...    ...   
submissiondate  56906    NaN   
starttime       56906    NaN   
endtime         56906    NaN   
subdate         56906    NaN   
enddate         56906    NaN   

                                                              top   freq  \
deviceid                                         4345ef039ad8e2b5   1209   
devicephonenum                                                     55777   
username                                                  collect  54311   
device_info     samsung|SM-T225|13|SurveyCTO Collect 2.80.2 (2...  36882   
duration                                                      201    190   
...                                                           ...    ...   
submissiondate                                                NaN    NaN   
starttime                                                     NaN    NaN   
endtime                                                       NaN    NaN   
subdate                                                       NaN    NaN   
enddate                                                       NaN    NaN   

                                         mean                  min  \
deviceid                                  NaN                  NaN   
devicephonenum                            NaN                  NaN   
username                                  NaN                  NaN   
device_info                               NaN                  NaN   
duration                                  NaN                  NaN   
...                                       ...                  ...   
submissiondate  2023-08-02 08:10:02.905159168  2023-07-22 12:42:41   
starttime       2023-08-02 04:46:56.341000704  2023-03-17 02:43:34   
endtime         2023-08-02 04:58:39.913945856  2023-07-22 08:27:26   
subdate         2023-08-01 16:50:58.854953984  2023-07-22 00:00:00   
enddate         2023-08-01 16:17:57.482163712  2023-07-22 00:00:00   

                                          25%                  50%  \
deviceid                                  NaN                  NaN   
devicephonenum                            NaN                  NaN   
username                                  NaN                  NaN   
device_info                               NaN                  NaN   
duration                                  NaN                  NaN   
...                                       ...                  ...   
submissiondate  2023-07-27 13:56:55.249999872  2023-08-01 17:32:54   
starttime                 2023-07-27 10:05:07  2023-08-01 16:17:46   
endtime            2023-07-27 10:13:17.500000  2023-08-01 16:23:44   
subdate                   2023-07-27 00:00:00  2023-08-01 00:00:00   
enddate                   2023-07-27 00:00:00  2023-08-01 00:00:00   

                                          75%                  max  std  
deviceid                                  NaN                  NaN  NaN  
devicephonenum                            NaN                  NaN  NaN  
username                                  NaN                  NaN  NaN  
device_info                               NaN                  NaN  NaN  
duration                                  NaN                  NaN  NaN  
...                                       ...                  ...  ...  
submissiondate  2023-08-07 15:58:51.750000128  2023-08-25 18:56:57  NaN  
starttime       2023-08-07 13:14:34.249999872  2023-08-25 18:33:15  NaN  
endtime            2023-08-07 13:19:46.500000  2023-08-25 18:40:53  NaN  
subdate                   2023-08-07 00:00:00  2023-08-25 00:00:00  NaN  
enddate                   2023-08-07 00:00:00  2023-08-25 00:00:00  NaN  

[450 rows x 11 columns]

In [8]:
data_entry_team_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 56906 entries, 0 to 56905
Columns: 456 entries, deviceid to v454
dtypes: category(407), datetime64[ns](5), float64(10), object(34)
memory usage: 43.9+ MB


In [9]:
# check columns that are entirely empty
empty_cols = [col for col in data_entry_team_1 if  data_entry_team_1[col].isna().all()]
print(empty_cols)

['kin1marks_p6_kin1p6_kin_q021p6_k', 'kin1marks_p6_kin1p6_kin_q061p6_k', 'kin1marks_p6_kin1p6_kin_q101p6_k', 'v453', 'kin1marks_p5_kin1p5_kin_q141p5_k', 'v454']


In [10]:
#drop empty cols
df_team_1 = data_entry_team_1.copy()
df_team_1 = df_team_1.drop(empty_cols,axis = 1)

# check columns that are entirely empty
empty_cols0 = [col for col in df_team_1 if  df_team_1[col].isna().all()]
print(empty_cols0)

[]


In [22]:
# let's check categorical columns
cat_cols = [col for col in df_team_1 if df_team_1[col].dtypes.name == "category"]
print(cat_cols)

cat_cols0 = df_team_1.select_dtypes(["category"]).columns
print(cat_cols0)

['spv', 'fo', 'eng_participation', 'p1_eng_q01_sub1', 'p1_eng_q01_sub2', 'p1_eng_q02_sub1', 'p1_eng_q02_sub2', 'p1_eng_q02_sub3', 'p1_eng_q03_sub1', 'p1_eng_q03_sub2', 'p1_eng_q03_sub3', 'p1_eng_q04_sub1', 'p1_eng_q04_sub2', 'p1_eng_q04_sub3', 'p1_eng_q05_sub1', 'p1_eng_q05_sub2', 'p1_eng_q05_sub3', 'p1_eng_q05_sub4', 'p1_eng_q05_sub5', 'p1_eng_q06_sub1', 'p1_eng_q06_sub2', 'p1_eng_q06_sub3', 'p1_eng_q07_sub1', 'p1_eng_q07_sub2', 'p1_eng_q07_sub3', 'p1_eng_q08_sub1', 'p1_eng_q08_sub2', 'p1_eng_q08_sub3', 'p2_eng_q01_sub1', 'p2_eng_q01_sub2', 'p2_eng_q01_sub3', 'p2_eng_q01_sub4', 'p2_eng_q02_sub1', 'p2_eng_q02_sub2', 'p2_eng_q02_sub3', 'p2_eng_q03_sub1', 'p2_eng_q04_sub1', 'p2_eng_q05_sub1', 'p2_eng_q06_sub1', 'p2_eng_q07_sub1', 'p2_eng_q07_sub2', 'p2_eng_q08_sub1', 'p2_eng_q08_sub2', 'p2_eng_q08_sub3', 'p2_eng_q09_sub1', 'p2_eng_q09_sub2', 'p2_eng_q09_sub3', 'p2_eng_q10_sub1', 'p2_eng_q10_sub2', 'p2_eng_q10_sub3', 'p2_eng_q10_sub4', 'p2_eng_q11_sub1', 'p3_eng_q01_sub1', 'p3_eng_q01_sub

we have 407 columns that are categorical, this makes sense where, except for svp and fo, the columns represent participation or marks received by students with 3 choices: correct, incorrect, or left blank

In [21]:
# let's visualize categorical columns
print(df_team_1[cat_cols])

                        spv                        fo eng_participation  \
0           Shema Christian        Ndagijimana Samuel               Yes   
1      NAHIMANA Reverien            NDABANANIYE David               Yes   
2           Shema Christian        Habonimana Gabriel               Yes   
3        NTAREMBA George             IRASUBIZA Zawadi               Yes   
4      NAHIMANA Reverien              KWIHANGANA Marc               Yes   
...                     ...                       ...               ...   
56901      Uwimana  Pauline  UTAZIRUBANDA NOEL Roudge               Yes   
56902  NAHIMANA Reverien               MWIZA Florence               Yes   
56903    NTAREMBA George              KWIHANGANA Marc               Yes   
56904       Shema Christian        Habonimana Gabriel               Yes   
56905  NAHIMANA Reverien              UWASE Francine                Yes   

      p1_eng_q01_sub1 p1_eng_q01_sub2 p1_eng_q02_sub1 p1_eng_q02_sub2  \
0                 NaN     

let's check for one of the categorical columns the choices available:

In [29]:
#print(df_team_1["p1_eng_q02_sub3"].cat.codes.unique())
#print(df_team_1["p1_eng_q02_sub3"].cat.categories.unique())
print(dict(enumerate(df_team_1["p1_eng_q02_sub3"].cat.categories)))
print(dict(enumerate(df_team_1["eng_participation"].cat.categories)))
print(dict(enumerate(df_team_1["spv"].cat.categories)))
print(dict(enumerate(df_team_1["fo"].cat.categories)))

{0: 'Left blank', 1: 'No', 2: 'Yes'}
{0: 'No', 1: 'Yes'}
{0: 'Shema Christian', 1: 'NAHIMANA Reverien   ', 2: 'Havugimana Saidi', 3: 'Uwimana  Pauline', 4: 'NTAREMBA George   ', 5: 'Hatangimana Gisele'}
{0: 'Rushambara Alexis', 1: 'NSHIMIYIMANA PASCAL', 2: 'Habonimana Gabriel', 3: 'Nirere Sandrine', 4: 'NIYIBIZI Leandre', 5: 'UWASE Francine\xa0', 6: 'Uwizeyimana Viateur', 7: 'Rusangwa Adolphe', 8: 'IRADUKUNDA GAHIRE ADOLPHE', 9: 'Ingabire Emelyne', 10: 'MUSHIMIRE Clarisse', 11: 'Mudahogora Placidie', 12: 'TUYISENGE ANICK', 13: 'Bagirishyaka Fulgence', 14: 'Fidele Iragena', 15: 'MUKANKOMEJE  Chantal', 16: 'NABAGIZE JUSTINE', 17: 'NTAWUSIGIRYAYO  Eric ', 18: 'IRAMBONA Valens', 19: 'Shingiro John', 20: 'UMUBYEYI MARIE GRACE', 21: 'MURAGIJIMANA Obadia', 22: 'MASENGESHO Samuel', 23: 'UMUBYEYI Claudine', 24: 'DUFATANYE Devota', 25: 'NSHIMIYIMANA Bernard', 26: 'MUKANSANGA Jacqueline', 27: 'UWIMANA Jeannette', 28: 'NAHAYO Jean Damascene', 29: 'DUSENGUMUREMYI Yvonne', 30: 'NSENGUMUREMYI Felix',

In [34]:
enumerators = dict(enumerate(df_team_1["fo"].cat.categories))
enumerators = pd.DataFrame(enumerators.items(),columns=["codes", "names"])
enumerators.head(20)

codes                      names
0       0          Rushambara Alexis
1       1        NSHIMIYIMANA PASCAL
2       2         Habonimana Gabriel
3       3            Nirere Sandrine
4       4           NIYIBIZI Leandre
5       5            UWASE Francine 
6       6        Uwizeyimana Viateur
7       7           Rusangwa Adolphe
8       8  IRADUKUNDA GAHIRE ADOLPHE
9       9           Ingabire Emelyne
10     10         MUSHIMIRE Clarisse
11     11        Mudahogora Placidie
12     12            TUYISENGE ANICK
13     13      Bagirishyaka Fulgence
14     14             Fidele Iragena
15     15       MUKANKOMEJE  Chantal
16     16           NABAGIZE JUSTINE
17     17      NTAWUSIGIRYAYO  Eric 
18     18            IRAMBONA Valens
19     19              Shingiro John

In [35]:
#we are going to convert categorical variables to their numeric values
df_team_1[cat_cols0] = df_team_1[cat_cols0].apply(lambda x: x.cat.codes)
#check the results
print(df_team_1.select_dtypes(["category"]).columns)
print(df_team_1["p1_eng_q02_sub3"].unique())
print(df_team_1["eng_participation"].unique())
print(df_team_1["spv"].unique())
print(df_team_1["fo"].unique())

Index([], dtype='object')
[-1  2  1  0]
[ 1  0 -1]
[0 1 4 2 5 3]
[110 109   2  89 103  99  36 112  13  98   5 137  40  48 119  26   8 140
 105  76  42  74  14  59  71 134  91 127  17   4  95  37 116 128  31  52
  61 121  51  96  77 102  54  55 122  97 113 139  21 136  35  62  90 108
  85 124  79  49  75 138  33  86  68  83 126  64  28  44  30  81   6  88
  23  41  11 107  67  65 120 100  15 123  34  10  43  47  57  94   9   3
  12  50  70  22  69 131 125 101  20  38  87  16  73  63  92   7 133  46
  45 104  80  78  84 114  32   0  72  66  53 117 130 118  24  93  18 106
  82  39  27  60 129 141  19  56  29  58 115   1 111 132  25 135]


In [37]:
df_team_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 56906 entries, 0 to 56905
Columns: 450 entries, deviceid to enddate
dtypes: datetime64[ns](5), float64(4), int16(1), int8(406), object(34)
memory usage: 41.2+ MB


we have successfully converted categorical variables to numeric variables

In [39]:
#check object type variables
object_cols = df_team_1.select_dtypes(["object"]).columns
print(object_cols)

Index(['deviceid', 'devicephonenum', 'username', 'device_info', 'duration',
       'caseid', 'district', 'sector', 'school', 'grade', 'student_code',
       'student_code_other', 'student_code_name', 'time_check_eng_1',
       'time_check_eng_2', 'time_check_eng_3', 'time_check_eng_4',
       'time_check_eng_5', 'time_check_eng_6', 'time_check_math_1',
       'time_check_math_2', 'time_check_math_3', 'time_check_math_4',
       'time_check_math_5', 'time_check_math_6', 'time_check_kiny_4',
       'time_check_kiny_5', 'time_check_kiny_6', 'time_check_kiny_1',
       'time_check_kiny_2', 'time_check_kiny_3', 'time_check_end',
       'instancename', 'key'],
      dtype='object')


In [40]:
# let's view the object type variables
df_team_1[object_cols].head()

deviceid devicephonenum  \
0  E0D3C54C-EFFE-4185-8F00-80D1CE7B79E6                  
1                      e9129293624d72e5                  
2                      d73df25fad2cc33b                  
3                      d6ee4c5c7759f9a6                  
4                      87c4ab681ae1d513                  

                          username  \
0                          collect   
1                          collect   
2                          collect   
3                          collect   
4  collect (not yet authenticated)   

                                         device_info duration caseid district  \
0  Apple|iPhone|15.7.5|SurveyCTO Collect 2.80 (27...      180          Nyanza   
1  samsung|SM-T225|13|SurveyCTO Collect 2.80.2 (2...      217         Musanze   
2  samsung|SM-T225|13|SurveyCTO Collect 2.80.2 (2...      575         Gakenke   
3  samsung|SM-T290|11|SurveyCTO Collect 2.80 (9b6...      401         Musanze   
4  samsung|SM-T505N|10|SurveyCTO Collect 2.80.2 (...      429          Nyanza   

     sector  school grade  ...     time_check_math_6     time_check_kiny_4  \
0    Busoro  210207    P3  ...   2023-Aug-7 15:09:59                         
1    Busogo  430114    P3  ...  2023-Jul-28 15:31:48                         
2  Kivuruga  420915    P5  ...  2023-Jul-24 09:12:33  2023-Jul-24 09:13:07   
3    Gacaca  430324    P6  ...  2023-Jul-31 10:55:21  2023-Jul-31 10:56:50   
4    Busoro  210209    P5  ...   2023-Aug-7 09:22:47   2023-Aug-7 09:22:50   

      time_check_kiny_5     time_check_kiny_6     time_check_kiny_1  \
0                                               2023-Aug-7 15:10:01   
1                                              2023-Jul-28 15:31:51   
2  2023-Jul-24 09:13:07  2023-Jul-24 09:15:24                         
3  2023-Jul-31 10:56:50  2023-Jul-31 10:56:50                         
4   2023-Aug-7 09:22:50   2023-Aug-7 09:25:09                         

      time_check_kiny_2     time_check_kiny_3        time_check_end  \
0   2023-Aug-7 15:10:01   2023-Aug-7 15:10:01   2023-Aug-7 15:10:01   
1  2023-Jul-28 15:31:51  2023-Jul-28 15:31:51  2023-Jul-28 15:31:51   
2                                              2023-Jul-24 09:15:24   
3                                              2023-Jul-31 10:59:55   
4                                               2023-Aug-7 09:25:09   

             instancename                                        key  
0  student_id210207190062  uuid:0000201e-8b84-40fa-8e37-05be1b8fac9b  
1  student_id430101190753  uuid:0000a616-72b4-485a-a7e3-e57ad1b48f16  
2                          uuid:00027ad6-86ea-49c2-aca1-eaf59f3eebc3  
3  student_id430311200702  uuid:00046f63-696f-44ba-b660-6395263d17eb  
4  student_id210209182748  uuid:00056471-339f-4339-8532-1944427947f9  

[5 rows x 34 columns]

In [42]:
#drop unuseful columns
df_team_1 = df_team_1.drop(['deviceid', 'devicephonenum', 'username', 'caseid','device_info','instancename', 'key'], axis=1)

In [43]:
#check object type variables again
object_cols0 = df_team_1.select_dtypes(["object"]).columns
print(object_cols0)

Index(['duration', 'district', 'sector', 'school', 'grade', 'student_code',
       'student_code_other', 'student_code_name', 'time_check_eng_1',
       'time_check_eng_2', 'time_check_eng_3', 'time_check_eng_4',
       'time_check_eng_5', 'time_check_eng_6', 'time_check_math_1',
       'time_check_math_2', 'time_check_math_3', 'time_check_math_4',
       'time_check_math_5', 'time_check_math_6', 'time_check_kiny_4',
       'time_check_kiny_5', 'time_check_kiny_6', 'time_check_kiny_1',
       'time_check_kiny_2', 'time_check_kiny_3', 'time_check_end'],
      dtype='object')


In [52]:
print(df_team_1["school"].unique())

['210207' '430114' '420915' '430324' '210209' '120604' '421102' '120405'
 '530425' '530801' '540504' '430907' '330213' '330208' '121201' '320509'
 '420542' '330802' '330309' '421007' '431206' '420411' '320115' '540117'
 '320507' '331127' '210208' '421502' '120805' '130713' '421208' '421611'
 '320208' '121203' '130313' '530717' '540613' '531107' '240203' '541211'
 '320611' '241006' '210713' '540208' '331213' '240404' '530908' '321210'
 '320401' '420413' '121432' '530418' '210408' '430701' '431108' '431009'
 '320201' '321108' '531311' '210407' '430517' '320601' '330308' '421808'
 '321001' '430906' '320207' '530818' '421404' '420101' '331212' '320802'
 '210908' '240709' '421509' '121509' '530426' '530503' '431104' '530819'
 '530511' '121431' '540609' '530104' '120308' '120506' '420513' '210901'
 '210125' '420311' '331205' '331216' '531208' '430903' '421207' '540610'
 '330214' '421614' '320409' '331133' '420502' '540210' '530702' '531312'
 '540215' '330816' '541201' '421402' '240804' '3312

In [45]:
print(df_team_1["student_code"].unique())
print(df_team_1["student_code_name"].unique())
print(df_team_1["student_code_other"].unique())
print(df_team_1["time_check_eng_1"].unique())
print(df_team_1["time_check_end"].unique())

['210207190062' '430101190753' '420908190716' ... '420302220018'
 '531103160670' '540457220024']
['' 'Abayo Ornella Bebeto' 'NISHIMWE HELENE' 'ISHIMWE OLIVIER'
 'SHIMWA BERGETH' 'NIYODUSHIMA ROSINE' 'Uwase Dusenge Pascaline'
 'Tuwombe Jean de Dieu' 'IRAKIZA DAMOUR' 'JWASE GIHOZO NADIA'
 'Uwimana Angelique' 'Niyomukiza Angelique' 'IRANKUNDA SABRINE'
 'Manzi fabrice' 'Nzayisenga leonard' 'UWIRAGIYE BLANDINE'
 'Kagabo Emmanuel' 'Ufashijwenimana Jonathan' 'Iragena Gisele'
 'Muragijimana Rosine' 'Tuyisenge Abrahamu' 'Niyoyavuze Jeremie'
 'Himbazwa Alpha Bryan' 'Iranzi Serine' 'AKINGENEYE VESTINE'
 'NISINGIZWE Vestine' 'Akimana Dathiva' 'Irakoze diedone'
 'Ingabire Claudine' 'GAD FAZILI EMANUEL' 'Dushimimana sandrine'
 'Kubwimana Adrien' 'Akariza jolly' 'UWINEZA Gloria' 'GIHOZO PATRICK'
 'NIYOGISUBIZO Fradja' 'Taremwa Caleb' "Nyiranibiringirwa Jeanne d'Arc"
 'Keza ange' 'Nyiramwiza Iradukunda' 'Imananiyokwizerwa Chadraque'
 'Uwacu Nicolas' 'Dukuzimana philimin' 'Uwizeyimana Kelvin'
 'Natasha

In [51]:
df_team_1[["student_code","student_code_other"]][df_team_1["student_code_other"]!="."]

student_code student_code_other
41     241006002001       241006002001
100    420311001002               -555
156    540712200041       540712200041
189    240505005003       240505005003
348    320208003002       320208003002
...             ...                ...
56207  320403230137       320403230137
56286  540609004001       540609004001
56389  210908190148       210908190148
56477  540712200267       540712200267
56828  430506181222       430506181222

[711 rows x 2 columns]

In [53]:
df_team_1["student_code_other"][41] == df_team_1["student_code"][41] 

True

In [58]:
df_team_1.loc[df_team_1["student_code_other"] == df_team_1["student_code"], "student_code_other"] = "."
print(df_team_1.student_code_other.unique())

['.' '-555' '555' '540712210384' '21108002025' '555555555555'
 '510712002010' '540712210425' '421807055001' '4305061550' '5707122003003'
 '4211020003025' '211001230088' '421102210191' '42111204027'
 '320028002001']


In [59]:
df_team_1.loc[df_team_1["student_code_other"] != ".", ["student_code","student_code_other"]]

student_code student_code_other
100    420311001002               -555
671    420311001003               -555
1268   421512190377                555
1443   420806001007                555
2433   421512190802                555
...             ...                ...
52350  420903001015                555
52646  421207003001                555
53804  421512190417                555
53823  421512002004                555
54817  421509005006                555

[64 rows x 2 columns]

In [60]:
df_team_1.loc[df_team_1["student_code_other"] != ".", "student_code_other"].unique()

array(['-555', '555', '540712210384', '21108002025', '555555555555',
       '510712002010', '540712210425', '421807055001', '4305061550',
       '5707122003003', '4211020003025', '211001230088', '421102210191',
       '42111204027', '320028002001'], dtype=object)

In [61]:
df_team_1.loc[df_team_1["student_code_other"].isin(['540712210384', '21108002025',
       '510712002010', '540712210425', '421807055001', '4305061550',
       '5707122003003', '4211020003025', '211001230088', '421102210191',
       '42111204027', '320028002001']), ["student_code","student_code_other"]]

student_code student_code_other
2800   540712210382       540712210384
5634   421108002025        21108002025
20610  540712002010       510712002010
21698  540712001110       540712210425
23344  421807005001       421807055001
24300  430506160550         4305061550
27585  540712003003      5707122003003
31098  421102003025      4211020003025
31101  211001002110       211001230088
33039  421102001100       421102210191
44755  421112004027        42111204027
46584  320208002001       320028002001

let's create a column that will help us track these inconsistencies

In [62]:
df_team_1["student_code_unsure"] = np.where(df_team_1["student_code_other"].isin(['540712210384', '21108002025',
       '510712002010', '540712210425', '421807055001', '4305061550',
       '5707122003003', '4211020003025', '211001230088', '421102210191',
       '42111204027', '320028002001']),"flagged","passed")


In [66]:
print(df_team_1["student_code_unsure"].value_counts())

passed     56894
flagged       12
Name: student_code_unsure, dtype: int64
